### PR TITLE
speedCommand

### PR DESCRIPTION
**Pull request by BakuDoriya-KatsZuku**

### CODE REVIEW

File: src/discord/commands/speed.ts [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+// import CliTable3 from "cli-table3";
+// import { ratingCmd } from "./rating";
+
+const INT_MAX = 2147483647;
+
+export const speedCmd: Command = {
+    info: new SlashCommandBuilder()
+        .setName("speed")
+        .setDescription("Fetch a user's speed of solving problems in contest"),
+        // .addBooleanOption(option => option
```
Comment (keerthivasansa) [inline src/discord/commands/speed.ts]: Remove these comments

File: src/discord/commands/speed.ts [hunk 2]
```diff
@@ -0,0 +1,101 @@
+import { AttachmentBuilder, SlashCommandBuilder } from "discord.js";
+import type { Command } from "../type";
+import { db } from "$db/index";
+import { CFApiFactory } from "$src/codeforces/client";
+import { CFLineChart } from "$src/graphs/line";
+import { EmbedBuilder } from "@discordjs/builders";
+// import { getRatingColor } from "$src/codeforces/range";
+// import CliTable3 from "cli-table3";
+// import { ratingCmd } from "./rating";
```
Comment (keerthivasansa) [inline src/discord/commands/speed.ts]: remove unnecessary imports

File: src/discord/commands/speed.ts [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+                prevContestID = allUserSubmissions[i].contestId;
+            }
+
+            const relativeTimeInMinutes = (relativeTimeInSeconds - prevRelativeTime) / 60;
+            if (!selectedData.has(problem.rating)) {
+                selectedData.set(problem.rating, [0, 0]);
+            }
+            const currentValue = selectedData.get(problem.rating)!;
+            currentValue[0] += relativeTimeInMinutes;
+            currentValue[1]++;
```
Comment (keerthivasansa) [inline src/discord/commands/speed.ts]: (optional) make the values clearer

current.sum  += relativeTimeInMinutes
curr.count++

is easier to read
similarly change the map definition too and the final map creation.

File: src/discord/commands/speed.ts [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+            prevRelativeTime = relativeTimeInSeconds;
+        }
+
+        const sortedByKey = new Map([...selectedData.entries()].sort((a, b) => a[0] - b[0]));
+        console.log(sortedByKey);
+        const finalData: Map<number, number> = new Map();
```
Comment (keerthivasansa) [inline src/discord/commands/speed.ts]: Better name - `ratingTimeMap`

File: src/discord/commands/speed.ts [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+                selectedData.set(problem.rating, [0, 0]);
+            }
+            const currentValue = selectedData.get(problem.rating)!;
+            currentValue[0] += relativeTimeInMinutes;
+            currentValue[1]++;
+
+            prevRelativeTime = relativeTimeInSeconds;
+        }
+
+        const sortedByKey = new Map([...selectedData.entries()].sort((a, b) => a[0] - b[0]));
```
Comment (keerthivasansa) [inline src/discord/commands/speed.ts]: Better name: `sortedByRating`

**Review by keerthivasansa — CHANGES_REQUESTED** [review]

**Comment by keerthivasansa** [comment]
Also resolve merge conflicts in `bun.lockb`

File: src/discord/commands/speed.ts (hunk 2 above)
Comment (BakuDoriya-KatsZuku) [reply src/discord/commands/speed.ts]: Done

File: src/discord/commands/stalk.ts [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+                colAligns: ['center', 'center', 'center'],
+                colWidths: [nameWidth, ratingWidth, submissionTimeWidth], // set the widths of each column (optional)
+            });
+
+            for (let i = start; i < end; i++) {
+                const problem = filteredSubmissions[i].problem;
+                const rating = problem.rating ? formatRating(problem.rating) : '?';
+                const name = wrapText(problem.name, nameWidth); // Wrap the text if it's too long
+
+                const timeAgo = (timeInS: number): string => {
```
Comment (keerthivasansa) [inline src/discord/commands/stalk.ts:128]: this should be a util function

File: src/discord/commands/stalk.ts [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+                } else {
+                    line += word + ' ';
+                }
+            }
+            result += line.trim(); // Add the last line
+            return result;
+        };
+
+        const getEmojiForRating = (rating: number): string => {
+            if (rating < 1200) return "🔘";
```
Comment (keerthivasansa) [inline src/discord/commands/stalk.ts:87]: maybe move this too to some tableUtils.ts within codeforces folder

File: src/discord/commands/stalk.ts [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+                return true;
+            }
+            return false;
+        });
+
+        const chunkSize = 10; // Show 10 submissions per page
+        const totalPages = Math.ceil(filteredSubmissions.length / chunkSize);
+
+        const wrapText = (text: string, maxWidth: number): string => {
+            const words = text.split(' ');
```
Comment (keerthivasansa) [inline src/discord/commands/stalk.ts:71]: this should be in utils/string.ts

File: src/discord/commands/stalk.ts (hunk 6 above)
Comment (keerthivasansa) [reply src/discord/commands/stalk.ts:128]: maybe in utils/date.ts

File: user_rating.json [hunk 9]
```diff
@@ -0,0 +1 @@
+[["2024-08-30T17:05:00.000Z",1664],["2024-09-20T16:35:00.000Z",1601],["2024-09-27T16:05:00.000Z",1609],["2024-09-29T17:35:00.000Z",1667],["2024-10-13T21:35:00.000Z",1557],["2024-10-14T16:35:00.000Z",1586],["2024-10-19T16:20:00.000Z",1632],["2024-10-20T11:05:00.000Z",1582],["2024-10-24T16:50:00.000Z",1577],["2024-10-26T16:35:00.000Z",1646]]
```
Comment (keerthivasansa) [inline user_rating.json]: delete this file

**Review by keerthivasansa — CHANGES_REQUESTED** [review]

**Review by keerthivasansa — APPROVED** [review]

**Pull request merged**